File: ts/206.reverse-linked-list.ts
/*
 * @lc app=leetcode id=206 lang=typescript
 *
 * [206] Reverse Linked List
 */

import ListNode from "./types/list-node";

// @lc code=start
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function reverseList(head: ListNode | null): ListNode | null {
  let prev = null;
  let curr = head;

  while (curr !== null) {
    const tmp = curr.next;
    curr.next = prev;
    prev = curr;
    curr = tmp;
  }

  return prev;
}

// function reverseList(head: ListNode | null): ListNode | null {
//   if (head === null || head.next === null) return head;

//   const prev = reverseList(head.next);

//   head.next.next = head;

//   head.next = null;

//   return prev;
// }

// @lc code=end

File: go/206.reverse-linked-list/206.reverse-linked-list.go
/*
 * @lc app=leetcode id=206 lang=golang
 *
 * [206] Reverse Linked List
 */
package leetcode

import (
	"github.com/maguroid/leetcode/go/structures"
)

type ListNode = structures.ListNode

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

/* iteration (faster than below) */
func reverseList(head *ListNode) (prev *ListNode) {
	for head != nil {
		head.Next, prev, head = prev, head, head.Next
	}

	return prev
}

// @lc code=end

/* iteration (standard) */
func reverseList2(head *ListNode) *ListNode {
	var prev *ListNode
	cur := head

	for cur != nil {
		tmp := cur.Next
		cur.Next = prev
		prev = cur
		cur = tmp
	}

	return prev
}

/* recursion */
func reverseList3(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	rev := reverseList3(head.Next)
	head.Next.Next = head
	head.Next = nil

	return rev
}
